pkg/network/master/subnets.go

```diff
@@ -235,6 +235,13 @@ func (master *OsdnMaster) handleAddOrUpdateSubnet(obj, _ interface{}, eventType
 
 	if err := common.ValidateHostSubnet(hs); err != nil {
 		utilruntime.HandleError(fmt.Errorf("Ignoring invalid HostSubnet %s: %v", common.HostSubnetToString(hs), err))
+		// If it was either the egressCIDR or egressIP fields that was invalid, then running an update will restore
+		// the hostsubnet object by wiping out the invalid egressCIDR/IP fields.
+		hs, err = master.networkClient.NetworkV1().HostSubnets().Update(context.TODO(), hs, metav1.UpdateOptions{})
```

Comment:
In general we do not undo the user's changes, even when they're wrong. So remove this part.

Reply:
> In general we do not undo the user's changes, even when they're wrong. So remove this part.

but we do right ? On startup in the sdn-controller, we reset invalid hostsubnet values.

https://github.com/openshift/sdn/blob/100a929557f10fe3a00707fb4b8e4f677944571f/pkg/network/master/subnets.go#L109

That's why I was trying to do something similar for the sdn-worker pods as well.

Reply:
@danwinship : so if I don't undo the user changes, how do I fix this bug ? The `reconcileHostSubnet` function does not automatically update the hostsubnet even if I clear the field. So I'd have to do the update here or add a new if condition inside `reconcileHostSubnet`

Reply:
Hm... that code is probably wrong. Basically it's not talking into account the fact that HostSubnet is half system-maintained (subnet allocations) and half user-maintained (EgressIPs). We don't want to be reverting mistakes to the user-maintained part.

`reconcileHostSubnet` is only there to deal with reconciliation of the system-maintained part; we have to do extra work to allocate subnets for F5 routers in one case.

In general, the behavior of openshift-sdn egress IP handling is not well-defined if the input egress IP data is invalid. The failing-to-start is clearly wrong, and we need to fix that, but other than that (and the fact that we shouldn't try to "fix" the HostSubnets ourselves), it doesn't much matter exactly what we do.

So, fix whatever is making it not start up, and ensure we're logging an error _somewhere_ about the bad IP, but otherwise, if we aren't going to take the time to write an admission controller, then there's not much we can do.

Reply:
ack thanks Dan, I also spoke with @squeed , he also suggests to fix the failing-on-startup part specifically if its failing due to invalid user-maintained parts (egress-CIDRs/IPs).

Reply:
right; if the system-maintained parts are invalid then that would mean that we couldn't set up the local network correctly anyway. But if the user-maintained/egressIP parts are invalid, we can at least get up and running, we just won't be able to set up the expected egress IPs

Reply:
cool, I've updated this patch. So when the `ValidateHostSubnet` function fails it returns an ErrorList with invalid fields. If those errors contain either `host`/`hostIP`/`subnet` I allow the cluster to fail, else we just ignore it - log it and move forward.

Casey, Dan : let me know if this lgty .